### PR TITLE
select: Add ScanRange to CSV&JSON

### LINKED ISSUE #14539: S3 Select: ScanRange does not seem to be suported
This was generally intuition, with select operations taking far longer than an equivalent call to s3. I then searched this repo, and couldn't find any instances of ScanRange, the param that s3 uses. It's also not mentioned in the docs, where other discussion of supported features lives.

## Expected Behavior
Given a call from Python such as:
```
response = self.s3_client.select_object_content(
                    Bucket=bucket,
                    Key=key,
                    ExpressionType='SQL',
                    Expression=f"select * from s3object s where s.pk = 'pk'",
                    InputSerialization={'CSV': {'FileHeaderInfo': 'Use', 'FieldDelimiter': '|'}},
                    OutputSerialization={'JSON': {}},
                    ScanRange={
                        'Start': 0,
                        'End': 1000
                    }
                )
```
I expect the call to return fairly quickly, even if the underlying file is several gigs

## Current Behavior
A call similar to that one (with python and minio running in docker) takes 16 seconds.

## Your Environment
* Version used (`minio --version`): docker minio/minio:RELEASE.2021-09-15T04-54-25Z

### PR DESCRIPTION
**Pull request by klauspost**
## Description

Implements https://docs.aws.amazon.com/AmazonS3/latest/API/API_SelectObjectContent.html#AmazonS3-SelectObjectContent-request-ScanRange

Fixes #14539

## How to test this PR?

See #14539

## Types of changes
- [x] New feature (non-breaking change which adds functionality)
- [x] Unit tests added/updated

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

File: internal/s3select/select.go [hunk 1]
```diff
@@ -298,9 +364,13 @@ func (s3Select *S3Select) getProgress() (bytesScanned, bytesProcessed int64) {
...
 // Currently CSV, JSON and Apache Parquet formats are supported.
 func (s3Select *S3Select) Open(getReader func(offset, length int64) (io.ReadCloser, error)) error {
+	offset, end, err := s3Select.ScanRange.StartLen()
+	if err != nil {
+		return err
+	}
 	switch s3Select.Input.format {
 	case csvFormat:
-		rc, err := getReader(0, -1)
+		rc, err := getReader(offset, end)
```
Comment (karlanke) [inline internal/s3select/select.go:373]: FYI I don't believe this matches the s3 logic - from https://docs.aws.amazon.com/AmazonS3/latest/userguide/selecting-content-from-objects.html: 
`A record that starts within the scan range specified but extends beyond the scan range will be processed by the query.`

Comment (karlanke) [reply internal/s3select/select.go:373]: Also just to confirm, is this going to work with search expressions? A surface read implies that the headers would no longer be available, but I haven't dug into how the reader is implemented.

Comment (klauspost) [reply internal/s3select/select.go:373]: @karlanke Both assumptions are correct and will not be changed.

**Pull request merged**